src/codebase.md

```diff
@@ -260,3 +260,41 @@ crates optimally for your semantic needs.
 | [ring](https://docs.rs/ring/)           | the leading crypto library         |
 | [nalgebra](https://docs.rs/nalgebra/)   | linear algebra                     |
 | [once_cell](https://docs.rs/once_cell/) | complex static data                |
+
+## How should I call C++ functions from Rust and vice-versa?
+
+Behind the scenes, Rust natively supports calling C functions using `extern "C"`,
+`#[repr(C)]` and `#[no_mangle]`. This is, of course, error prone and involves
+lots of `unsafe` code. So don't do that: there are better ways.
+
+Your initial decision should be a matter of policy. Choose between:
+
+* I want to manually specify the cross-language interface in order to keep
+  a degree of isolation and to avoid proliferation of uncontrolled
+  cross-language boundaries. Or:
+* I want to generate comprehensive bindings automatically from existing C++
+  headers to allow arbitrary calls from Rust to C++.
+
+If you choose the former approach, the modern standard for C++ interoperability
+is [cxx](https://cxx.rs). You specify cross-language interfaces in a
+Rust-like interface definition language (within your Rust file). Both Rust and
+C++ bindings code is generated, to make the interface natural in each language. For
```

Comment:
"bindings code is generated"

Maybe just "bindings are generated"?

"to make the interface natural in each language"

Not sure this is the actual reason for generating bindings in both languages? I'd have said you need bindings simply to be able to call back and forth at all (assuming you don't want to use unsafe code).

All in all, how about:

"Both Rust and C++ bindings can be generated, allowing Rust to call C++ and vice versa."

Reply:
Hmm, I think that misses the point I'm trying to make.

The reason the `cxx` is so pleasant to use is because there is bindings code generated on both sides. A call from Rust to C++ involves marshaling Rust data to a lowest common denominator C ABI format, then unmarshaling back on the C++ side. The same applies in reverse.

I've reworded, and I may add a few diagrams.

Reply:
> Hmm, I think that misses the point I'm trying to make.
> 
> The reason the `cxx` is so pleasant to use is because there is bindings code generated on both sides. A call from Rust to C++ involves marshaling Rust data to a lowest common denominator C ABI format, then unmarshaling back on the C++ side. The same applies in reverse.

Ah, got it. Yes, that's something different than what the wording I was suggesting expresses.

> I've reworded,

Hm, I don't see a commit... did you just mean to say "I'll reword"?

> and I may add a few diagrams.

Reply:
"Done" locally, but not pushed... until now. That seems to be the most natural way to handle github review comments.

Reply:
Ah. Should have waited a bit longer.

Rewording looks good to me!